apps/api_web/lib/api_web/event_stream.ex

```diff
@@ -37,50 +38,54 @@ defmodule ApiWeb.EventStream do
     ensure_timer(%__MODULE__{conn: conn, pid: pid, timeout: timeout})
   end
 
-  @spec hibernate_loop(state) :: no_return | {:error, term}
+  @spec hibernate_loop(state) :: Plug.Conn.t()
   def hibernate_loop(state) do
     case receive_result(state) do
-      {:ok, state} ->
+      {:continue, state} ->
         :proc_lib.hibernate(__MODULE__, :hibernate_loop, [state])
 
-      error ->
+      {:close, conn} ->
         Supervisor.server_unsubscribe(state.pid)
-        error
+        conn
     end
   end
 
-  @spec receive_result(state) :: {:ok, state} | {:error, term}
-  def receive_result(state) do
+  @spec receive_result(state) :: {:continue, state} | {:close, Plug.Conn.t()} | {:error, term}
+  def receive_result(%{conn: conn, pid: pid} = state) do
     receive do
       {:events, events} ->
-        chunks =
-          for {type, item} <- events do
-            ["event: ", type, "\ndata: ", item, "\n\n"]
-          end
+        chunks = for {type, item} <- events, do: ["event: ", type, "\ndata: ", item, "\n\n"]
+        continue(state, chunks)
 
-        update_state(state, chunk(state.conn, chunks))
+      :timeout ->
+        continue(state, ": keep-alive\n")
 
       {:error, rendered} when is_list(rendered) ->
-        {:error, chunk(state.conn, ["event: error\ndata: ", rendered, "\n\n"])}
+        close(state, ["event: error\ndata: ", rendered, "\n\n"])
 
-      :timeout ->
-        update_state(state, chunk(state.conn, ": keep-alive\n"))
+      {:DOWN, _ref, :process, ^pid, reason} ->
+        if reason in [:normal, :shutdown] or match?({:shutdown, _}, reason),
```

Comment:
Maybe better as a new function w/ pattern matching?

Reply:
I wanted to avoid repetition of `close(state)`, which I don't think extracting this to a function would accomplish.

Reply:
It would also place the `close` call at a different level of abstraction than the other `continue`/`close` calls, and I think having them at the same level is better for readability.